### ISSUE TITLE
Cannot type narrow APIApplicationCommandInteraction

### ISSUE THREAD

**Issue by andre4ik3** [discordjs/discord-api-types#340]
**Please describe the problem you are having in as much detail as possible:**

When trying to type narrow interaction object, it is impossible to type narrow it to chat input. I looked at interaction types and I can't work out why it is like that, so I'm asking for help. Maybe it is my issue, but I do not think so.

**Include a reproducible code sample here, if possible:**

I have function that takes interaction as APIChatInputApplicationCommandInteraction. 

```ts
// Say you have an interaction and want to check if it's chat input (slash command)
if (interaction.type === 2 && interaction.data.type === 1) {
  // here interaction should definitely be chat input
  // but it's not, it still APIApplicationCommandInteraction
  // and can't assign to APIChatInputApplicationCommandInteraction
  await this.execute(interaction); // Type "APIApplicationCommandInteraction" not assignable to "APIChatInputApplicationCommandInteraction"
```

**Further details:**

<img width="780" alt="CleanShot 2022-02-18 at 20 43 04@2x" src="https://user-images.githubusercontent.com/62390580/154735711-cc42c586-2761-4d9c-92e3-4e51849d752b.png">

- Runtime:
  - Node.js version: v17.5.0
- Priority this issue should have – please be realistic and elaborate if possible:
  Medium because it is a bug but can be bypass, e.g. by `interaction as APIChatInputApplicationCommandInteraction`

**Comment by vladfrangu** [comment]
CNR 
![image](https://user-images.githubusercontent.com/17960496/154737618-9eb57f1e-968e-447f-8553-ac5b19c7f9a6.png)

**Comment by suneettipirneni** [comment]
You can only narrow one property at time. In the first condition of your if statement typescript sees you're trying to narrow the interaction, not the data type *of that* interaction, only the the first condition is taken into consideration since the other is another property.

[Playground demo](https://www.typescriptlang.org/play?#code/KYOwrgtgBAKg9gBxgTwcAzlA3gKCvqAeRGABo8CYB3Ocg2ACwCdhgcBfHHUSKAZTAAjFGky56xMhXzVa0xiymccASxAAXYEwBmAQwDGwKAEFs89amAAuWIhEYAdJLoFtcODfTqmagOYBueXQhGwFBY0DlNU0dAyNjACYzegs0G3gkS3QnEhd8Nw8oLx8QAKCQ-iEAIUiuaK09Qygq5IJU61tM0QdZPKhBXSYbcAhBLVrVDQa4qABhVvx29LssnuZWPoGALxtBFV9oifbOqABeEygAHxMk65br2bqp2Kaw03E2y1Che2zJQPoBU83j8E3qLyMYRaH0WX0qwlWsgBBAGQygIzGTDBz0akKE8xhUCW8N+a0UyPw212+0OHC4xzCZ3hpmuUKu8NmgRwABNgPoADaDIz6OAgLxQOCCABWywQXJU2igAApJVKHMdTprOqTJFAAGR6iXShzBQTqyxnLVhHUkACUCyNUpwAHpnQQAHoAfg4QA)

**Comment by andre4ik3** [comment]
So how would I do it? I thought two separate if statements but that is weird and doesn't work anyway

<img width="388" alt="CleanShot 2022-02-18 at 23 10 32@2x" src="https://user-images.githubusercontent.com/62390580/154754577-7f46e333-93a1-454e-87ec-c740d9ae9852.png">

**Comment by vladfrangu** [comment]
For now, either cast or wait till we implement typeguards 😅

**Comment by andre4ik3** [comment]
Ok.

**Comment by Loremente** [comment]
SOS

**Comment by andre4ik3** [comment]
I was able to work around this issue by... creating my own framework based on discord-api-types that has classes around interactions with type guards lol.

I don't think there is anything possible for discord-api-types, as it's a type library, and type guards are functions that do checks at runtime, which would make interactions classes, which would take discord-api-types out of devDependencies and bloat compiled code...

So yeah, the only thing possible while keeping discord-api-types purely a typings library would be to cast. 😕

**Comment by andre4ik3** [comment]
Also interaction type guards just got merged in to 0.28.0 in utils, so I guess this is solved :)
